src/libs/compound-v3/service.ts

```diff
@@ -0,0 +1,348 @@
+  assets: {
+    asset: common.Token;
+    priceFeed: string;
+    borrowCollateralFactor: string;
+    liquidateCollateralFactor: string;
+  }[];
+}
+
+export class Service extends logics.compoundv3.Service {
+  static provider?: providers.Provider;
+
+  readonly blockTag?: providers.BlockTag;
+
+  constructor(chainId: number, blockTag?: providers.BlockTag) {
+    super(chainId, Service.provider);
+    this.blockTag = blockTag;
+  }
+
+  private _marketMap: Record<string, Market> = {};
+
+  async getMarket(marketId: string) {
+    if (!this._marketMap[marketId]) {
+      const cometAddress = logics.compoundv3.getMarket(this.chainId, marketId).cometAddress;
+      const iface = logics.compoundv3.Comet__factory.createInterface();
+
+      let baseTokenAddress: string;
+      let baseTokenPriceFeed: string;
+      let numAssets: number;
+      let utilization: BigNumber;
+      {
+        const calls: common.Multicall2.CallStruct[] = [
+          {
+            target: cometAddress,
+            callData: iface.encodeFunctionData('baseToken'),
+          },
+          {
+            target: cometAddress,
+            callData: iface.encodeFunctionData('baseTokenPriceFeed'),
+          },
+          {
+            target: cometAddress,
+            callData: iface.encodeFunctionData('numAssets'),
+          },
+          {
+            target: cometAddress,
+            callData: iface.encodeFunctionData('getUtilization'),
+          },
+        ];
+        const { returnData } = await this.multicall2.callStatic.aggregate(calls, { blockTag: this.blockTag });
+
+        [baseTokenAddress] = iface.decodeFunctionResult('baseToken', returnData[0]);
+        [baseTokenPriceFeed] = iface.decodeFunctionResult('baseTokenPriceFeed', returnData[1]);
+        [numAssets] = iface.decodeFunctionResult('numAssets', returnData[2]);
+        [utilization] = iface.decodeFunctionResult('getUtilization', returnData[3]);
+      }
+
+      const assetAddresses: string[] = [];
+      const assetPriceFeeds: string[] = [];
+      const borrowCollateralFactors: string[] = [];
+      const liquidateCollateralFactors: string[] = [];
+      {
+        const calls: common.Multicall2.CallStruct[] = [];
+        for (let i = 0; i < numAssets; i++) {
+          calls.push({ target: cometAddress, callData: iface.encodeFunctionData('getAssetInfo', [i]) });
+        }
+        const { returnData } = await this.multicall2.callStatic.aggregate(calls, { blockTag: this.blockTag });
+
+        for (let i = 0; i < numAssets; i++) {
+          const [{ asset, priceFeed, borrowCollateralFactor, liquidateCollateralFactor }] = iface.decodeFunctionResult(
+            'getAssetInfo',
+            returnData[i]
+          );
+          assetAddresses.push(asset);
+          assetPriceFeeds.push(priceFeed);
+          borrowCollateralFactors.push(common.toBigUnit(borrowCollateralFactor, 18));
+          liquidateCollateralFactors.push(common.toBigUnit(liquidateCollateralFactor, 18));
```

Comment:
Are those decimals related to the collateral decimal?

Reply:
No~